### ISSUE TITLE
VRM Importing

### ISSUE THREAD

**Issue by ProbablePrime** [Yellow-Dog-Man/Resonite-Issues#7]
### Is your feature request related to a problem? Please describe.

VRM is an "all in one" file format for 3D Avatars that many platforms already support. There are lots of VRM tools and avatars out there.

VRM imports into Resonite are currently not fully supported. The underlying GLTF data can be looked at but other VRM features are ignored.

### Describe the solution you'd like

I'd like Resonite to support the VRM standard for import/setup of avatars as closely as it can.

### Describe alternatives you've considered

Extracting the GLTF data from a VRM is possible but much of the meta-data is lost.

### Additional Context

I am pre-seeding the issue repository with commonly requested items to prevent those issues from being made as duplicates once open.

**Comment by Casuallynoted** [comment]
I think it would be worth linking this discussion from the Discord as well:

https://discord.com/channels/1040316820650991766/1158258132435337266

It discusses the use of VRM metadata and licensing restrictions, and various ways that Resonite may be able to go about those in a way that is respectful to creators.

**Comment by kazu0617** [comment]
From the perspective of writing a guide article for Japanese people at Resonite and creating conversion software, I will write down the contents that may be necessary.
## Links
What is VRM? : https://vrm.dev/
Article on how to put in as much as I could find about VRM in Japan (I share this article to have it put in. Yes, of course most of the data is missing...): https://sharedx.notion.site/VRM-d93c6e3ae2f647e0956054efff1d20b9?pvs=4

## About the version
VRM is available in 0.x and 1.0. You can tell the difference between them by checking the contents of gltf and see if the add-on you have is "vrm" or "VRMC_vrm-1.0".

## About data
As Prime said, it is an extension of gltf. So I think it is *appropriate to load [.vrm] as gltf2.0 as the first landing point*.
Of course, it would be nice if all or a series of data could be loaded properly (e.g., a set of shape keys could be baked and added as expressions, so that minimal modifications are required after installation, or the avatar setting tool could appear in the position where it was set after installation), (It would be great if SpringBone's settings could be roughly formatted as text (even if they are not compatible, it would be given to us as data at first).

**Comment by bdunderscore** [comment]
VRM also contains metadata relevant to avatar behavior, such as viewpoint positions, facial expressions, springbones, material configuration, etc…

**Comment by Geenz** [comment]
> As noted above- the only functional difference between VRM, and GLTF is the existence of licensing meta-data. As such, this issue is blocked by the implementation of a licensing system, e.g. #626.

Licensing metadata is different in this case, and may not be trivial to really respect beyond the user acting in good faith.  Basically our concept of a license is not likely to fully align with VRM's concept of a license.

**Comment by aaronfranke** [comment]
As mentioned above, VRM is not just a monolithic format, it is composed of many pieces, and all these pieces do a lot more than just licensing. Each `.vrm` file is really a `.glb` file with VRM-specific glTF extensions, including:

- Node constraints: Allows controlling the notation of one node/bone based on another. Can be used to implement things like digitigrade legs, taur legs, and multi-arm characters, but could also be used outside of characters.
- Spring bones: Allows for nodes/bones to spring/jiggle. Can be used to implement things like floppy ears or a dynamic tail on a furry character, but could also be used outside of characters.
- Toon materials: Allows for a less realistic toon look for meshes, usually characters, but could be other objects too.
- Animations: Allows expressions and animations to be defined on an avatar.
- Bone mapping: The base VRM extension requires a humanoid bone map, similar to Unity's bone map menu.
- First person: The base VRM extension allows defining things to hide when in first person, like hiding the mesh of the head/glasses/etc when in VR but still allowing you to look down and see your body.
- Meta: The base VRM extension defines metadata like licensing information. Also note, there is some overlap with Khronos's own KHR_xmp_json_ld extension for copyright metadata (see [spec](https://github.com/KhronosGroup/glTF/tree/main/extensions/2.0/Khronos/KHR_xmp_json_ld) and [Godot implementation](https://godotengine.org/asset-library/asset/2309)).

Therefore, VRM is not just one monolithic thing, it is a standard built out of many pieces. One does not necessarily have to implement the entirety of VRM. You can for example take the extension for VRM constraints and use it outside of VRM in a regular .glb file, as I did [here](https://www.youtube.com/watch?v=vI0baYrC5Co). It makes sense to provide these general tools for any glTF model. The complete list of VRM extensions can be found here: https://github.com/vrm-c/vrm-specification/tree/master/specification

This issue is highly related to https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2778, since VRM is based on glTF. The same codebase can be used to import these files, optionally mixing and matching VRM and non-VRM features.

**Comment by ProbablePrime** [comment]
I looked into this a little bit due to some business requirements in a nearby area.

There's a lot of conflation that exists here between a variety of items, but we could split those up a bit. A suitable split might be:
1. Fixing up any remaining GLTF issues, such as #155 
2. Hooking up Resonite to just import vrm as GLTF without processing the meta-data at all
3. Incorporating any suitable metadata on import that might be helpful
	- Hierarchy issues can be helped. if the meta data tells us which bones to do (based on bone mapping above)
	- Conversion between spring bones and dynamic bones
	- View point
	- hand location etc
4. After we have license system, honoring the meta-data of the VRM license data

Splitting it up a bit like that, may help to get some blockers moved aside.

**Comment by ko-tengu** [comment]
In the most basic sense, Resonite already supports VRM so long as you first rename the files to .GLB and I've done this dozens of times in the past. If you want to split it up and add format support piecemeal it can be done at the very most basic level by simply adding VRM to the glTF import extension list.

Having support for the most relevant metadata components would be preferable long term however.

**Comment by TumnusB** [comment]
A member of JP Community made the following 

https://github.com/kazu0617/vrmtoglb_autoconvert

Which auto renames bones and some other stuff that helps with importing a VRM into Resonite

**Comment by april83c** [comment]
any updates on this? still my #1 showstopper

**Comment by kazu0617** [comment]
@april83c 
The Japanese community is increasingly converting VRM -> VRChat system -> Resonite rather than directly converting VRM.

They also released Modular Avatar - Resonite Support as a way to bring VRChat avatars directly into Resonite.
Although Poiyomi does not support (So far. unless someone wants to contribute.) movements, or some movements such as Constraint/Cloth, it is now very easy to convert.
Doc: https://modular-avatar.nadena.dev/docs/experimental-features/resonite-support
Blog(need translate): https://dinosaur-fossil.hatenablog.com/entry/2025/05/28/125047

**Comment by TahvoDev** [comment]
1. VRM may be directly created from multiple applications, including VRoid Studio and even Blender (with an add-on). 
2. Some people (especially vtubers) may already have a VRM file for other applications (VSeeFace, Warudo, LIV). Directly importing to a fully usable avatar would make getting started easier.

**Comment by april83c** [comment]
> Don't you just rename the .vrm to .glb and then import it like a normal avatar?
> 
> But I guess something that can read the "Spring Bones' and stuff from the VRM extra data to help import the model would be cool?
> But most times VRM models I've seen are just people taking the VRchat unity project and just porting it over into a VRM.   So you'd already have the RAW FBX to start with.  Why convert from one format to another where stuff is removed during creation of the VRM to reduce file sizes...

...no?

The entire point of the existence of this issue is that that *isn't* the case.

If you import a VRM by renaming it into a GLB you get a fully white untextured mesh that you have to manually set up as an avatar.

Properly importing a VRM would mean you drag the VRM file in and have a fully working avatar, with properly mapped bones, a properly set eye viewpoint, properly configured physics bones, proper textures, materials/shaders configured as they should be to make the avatar look as intended - everything already done.

VRM is not a format for storing 3D models, VRM is a format for storing avatars to be used in virtual worlds

**Comment by blhsrwznrghfzpr** [comment]
> Don't you just rename the .vrm to .glb and then import it like a normal avatar?

At least for now, avatars created with VRoid Studio in VRM 1.0 cannot be imported just by renaming `.vrm` to `.glb`.

**Comment by kazu0617** [comment]
> > Don't you just rename the .vrm to .glb and then import it like a normal avatar?
> 
> At least for now, avatars created with VRoid Studio in VRM 1.0 cannot be imported just by renaming `.vrm` to `.glb`.

VRM1.0 uses Draco compression in meshes, that doesn't support resonite now.
https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/155
